### PR TITLE
support for jax objectives

### PR DESCRIPTION
**Pull request by FFroehlich**
supports combination of arbitrary pypesto objectives with jax
uses host callbacks (experimental feature) to circumvent XLA computation
caching borrowed from aesara code, which we could deprecate now

~still needs some documentation.~

### CODE REVIEW

File: pypesto/objective/jax/base.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+
+try:
+    import jax
+    import jax.experimental.host_callback as hcb
+    import jax.numpy as jnp
+    from jax import custom_jvp, grad
+except ImportError:
+    raise ImportError(
+        "Using a jax objective requires an installation of "
+        "the python package jax. Please install aesara via "
```
Comment (yannikschaelte) [inline pypesto/objective/jax/base.py:26]:
```suggestion
        "the python package jax. Please install jax via "
```

Comment (yannikschaelte) [inline pypesto/objective/jax/base.py:26]:
```suggestion
        "the python package jax. Please install jax via "
```

File: pypesto/objective/jax/base.py [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+# https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html
+
+
+@partial(custom_jvp, nondiff_argnums=(0,))
+def _device_fun(obj: 'JaxObjective', x: jnp.DeviceArray):
+    """Jax compatible objective function execution using host callback.
+
+    This function does not actually call the underlying objective function,
+    but instead extracts cached return values. Thus it must only be called
+    from within obj.call_unprocessed and obj.cached_base_ret must be populated.
```
Comment (yannikschaelte) [inline pypesto/objective/jax/base.py:40]:
```suggestion
    from within obj.call_unprocessed, and obj.cached_base_ret must be populated.
```

File: pypesto/objective/jax/base.py [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+    from jax import custom_jvp, grad
+except ImportError:
+    raise ImportError(
+        "Using a jax objective requires an installation of "
+        "the python package jax. Please install aesara via "
+        "`pip install jax jaxlib`."
+    )
+
+# jax compatible (jittable) objective function using host callback, see
+# https://jax.readthedocs.io/en/latest/jax.experimental.host_callback.html
```
Comment (yannikschaelte) [inline pypesto/objective/jax/base.py:31]: :+1:

File: pypesto/objective/jax/base.py [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+class JaxObjective(ObjectiveBase):
+    """Objective function that combines pypesto objectives with jax functions.
+
+    Parameters
+    ----------
+    objective:
+        pyPESTO objective
+    jax_fun:
+        jax function (not jitted) that computes input to the pyPESTO objective
```
Comment (yannikschaelte) [inline pypesto/objective/jax/base.py:154]: So `objective(jax_fun(x))` will be executed? Could be stated more clearly.

**Review by yannikschaelte — APPROVED** [review]
nice!

**Comment by yannikschaelte** [comment]
(An application example demonstrating the usefulness could ofc be nice, but not a necessity)

**Comment by FFroehlich** [comment]
> (An application example demonstrating the usefulness could ofc be nice, but not a necessity)

TBD

**Pull request merged**